src/cc/lib/py_graph.cc

```diff
@@ -925,6 +926,123 @@ int32_t RandomWalk(PyGraph *py_graph, int64_t seed, float p, float q, NodeID def
+                          float *out_weights)
+{
+    if (py_graph->graph == nullptr)
+    {
+        RAW_LOG_ERROR("Internal graph is not initialized");
+        return 1;
+    }
+
+    const float alpha_eps = alpha * eps;
+    std::vector<NodeID> q;
+
+    std::vector<NodeID> neighbors;
+    std::vector<float> weights;
+    std::vector<Type> types;
+
+    absl::flat_hash_map<NodeID, float> p;
+    absl::flat_hash_map<NodeID, float> r;
+    using WN = std::pair<float, NodeID>;
+    std::priority_queue<WN, std::vector<WN>, std::greater<WN>> pq;
+    std::vector<uint64_t> neighbor_counts;
+    for (size_t node_index = 0; node_index < int_node_ids_size; ++node_index)
+    {
+        p.clear();
+        r.clear();
+        const auto inode = in_node_ids[node_index];
+        r[inode] = alpha;
+        q.emplace_back(inode);
+        while (!q.empty())
+        {
+            auto unode = q.back();
+            q.pop_back();
+            float res = 0;
+            auto r_unode = r.find(unode);
+            if (r_unode != std::end(r))
+            {
+                res = r_unode->second;
+            }
+            p[unode] += res;
+            r[unode] = 0;
+
+            neighbors.resize(0);
+            weights.resize(0);
+            types.resize(0);
+            uint64_t curr_count = 0;
+            GetNeighborsInternal(py_graph, &unode, 1, in_edge_types, in_edge_types_size, &curr_count, neighbors, types,
+                                 weights);
+            neighbor_counts.resize(neighbors.size());
+            NeighborCount(py_graph, neighbors.data(), neighbors.size(), in_edge_types, in_edge_types_size,
+                          neighbor_counts.data());
+            const float _val = (1 - alpha) * res / neighbors.size();
+            for (size_t nb_index = 0; nb_index < neighbors.size(); ++nb_index)
+            {
+                auto vnode = neighbors[nb_index];
+                r[vnode] += _val;
+
+                float res_vnode = 0;
+                auto r_vnode = r.find(vnode);
+                if (r_vnode != std::end(r))
+                {
+                    res_vnode = r_vnode->second;
+                }
+
+                if (res_vnode >= alpha_eps * neighbor_counts[nb_index])
+                {
+                    auto f = std::find(std::begin(q), std::end(q), vnode);
+                    if (f == std::end(q))
+                    {
+                        q.emplace_back(vnode);
+                    }
+                }
+            }
+        }
+
+        for (auto kv : p)
+        {
+            if (pq.size() < count)
+            {
+                pq.emplace(kv.second, kv.first);
+            }
+            else
+            {
+                const auto &top = pq.top();
+                if (top.first > kv.second)
+                {
+                    continue;
+                }
+                pq.pop();
+                pq.emplace(kv.second, kv.first);
```

Comment:
Should we add this to benchmarks? Does it scale?

Reply:
PPR sampling is similar to random walk. It is implemented at the library interface level and we can't include it in existing benchmarks. Scaling similarly depends on input parameters, alpha and eps.